### PR TITLE
feat(browse classes within program): page and api

### PR DESCRIPTION
**Pull request by MatoPlus**
### Notion ticket link

[Browse all classes within a program page](https://www.notion.so/uwblueprintexecs/Browse-all-classes-within-a-program-page-5cb3fc19a829426380642d1d2799fd8c)

### Implementation description

Teacher View:

https://user-images.githubusercontent.com/44826218/142775929-c30b1698-9b2b-4092-8e7c-47563c326a03.mp4

Admin View:

https://user-images.githubusercontent.com/44826218/142775935-ced47d78-8136-4687-ac81-534d0053595a.mp4

- Add new page:  /admin/program/[pid].tsx which reads the program and all it's classes and displays it as per designed
- Add generic loading and error pages for cases where api fails (eg: when program is not found or still loading) and update all loading and error cases to use the generic components
- new archive api route, takes a program id and updates the program and all its classes
- add generic button for consistent theming - TODO: fix redundancy in past use of buttons
- add generic admin modal seen in demo - @GodGreg this is important for your ticket!
- fix fetcher to throw error so useSWR returns a working error object
- add use programs and update program hooks/calls

### Steps to test

NOTE: There's way to many cases to list out - might be a good idea to blitz

### What should reviewers focus on?

- Dev: code style
- Design: is the new generic error page okay?
- Design: is generic Modal okay?
- Design: modal and toast messages

### Checklist

-   [X] My PR name is descriptive and in imperative tense
-   [X] I have run the linter
-   [X] I have requested a review from the PL, and/or other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

**Comment by MatoPlus** [comment]
Case for archive program:

https://user-images.githubusercontent.com/44826218/142776437-ba728c7b-9b81-4121-9696-2b1ec09b3365.mp4

File: src/components/AdminError.tsx [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import React from "react";
+import Wrapper from "@components/AdminWrapper";
+import { Error } from "@components/Error";
+
+export type AdminErrorProps = {
+    cause: string;
```
Comment (MatoPlus) [inline src/components/AdminError.tsx:9]: add session prop here.

File: src/components/admin/ProgramViewInfoCard.tsx [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+import { AdminModal } from "./AdminModal";
+import { roles } from "@prisma/client";
+
+export type ProgramViewInfoCard = {
+    cardInfo: ProgramCardInfo;
+    role: roles;
+};
+
+/**
+ * Admin view class card component used in the admin class details page
```
Comment (MatoPlus) [inline src/components/admin/ProgramViewInfoCard.tsx]: comment update

File: src/components/admin/ProgramClassInfoCard.tsx [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+import { AdminModal } from "./AdminModal";
+
+export type ProgramClassInfoCard = {
+    cardInfo: ClassCardInfo;
+    role: roles;
+    mutateClasses: (data?: any, shouldRevalidate?: boolean) => Promise<any>;
+};
+
+/**
+ * Admin view class card component used in the admin class details page
```
Comment (MatoPlus) [inline src/components/admin/ProgramClassInfoCard.tsx]: change

**Review by bekahma — APPROVED** [review]
lgtm 👍

File: utils/updateProgramArchive.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+/**
+ * updateProgramArchive updates a program for whether or not it is archived with admin permission
+ * @param id id of program
+ * @param isArchive whether or not the program should be archived
+ * @returns updated program
+ */
+export async function updateProgramArchive(
+    id: number,
+    isArchive: boolean,
+    /* eslint-disable @typescript-eslint/no-explicit-any */
```
Comment (MatoPlus) [inline utils/updateProgramArchive.ts]:
```suggestion
```

**Pull request merged**